core/stdlib/std.ncl

```diff
@@ -1371,39 +1364,62 @@
 
               # Examples
 
+              This example is illustrative. In practice, returning `'Error {..}`
+              should always be preferred over manipulating the label directly
+              whenever possible.
+
               ```nickel
-              let AlwaysFailWithNotes = std.contract.delayed (fun label _ =>
+              let AlwaysFailWithNotes = std.contract.custom (fun label _ =>
                 label
                 |> std.contract.label.append_note "This is note 1"
                 |> std.contract.label.append_note "This is note 2"
                 |> std.contract.blame
               )
               in
+
               null | AlwaysFailWithNotes
+              ```
             "%
           = fun note label => %label/append_note% note label,
       },
 
     apply
       | doc m%"
-          Applies a contract to a label and a value.
+          Applies a contract to a label and a value. Either aborts the execution
+          with a broken contract error if the contract fails, or proceed with
+          the evaluation of the value.
+
+          Using a contract annotation, such as in `value | Contract`, eventually
+          translates to a call such as `std.contract.apply Contract label value`
+          where `label` is inserted by the Nickel interpreter.
 
           Type: `Contract -> Label -> Dyn -> Dyn`
           (for technical reasons, this function isn't actually statically typed)
 
-          Nickel supports user-defined contracts defined as predicates, as
-          general functions, as records, etc. The interpreter also performs additional
-          bookkeeping for error reporting when applying a contract in an
-          expression `value | Contract`. You should thus not use standard
+          Although contracts used to be naked functions a long time ago,
```

Comment:
I think there's too much discussion of the old state of affairs. The changes should be documented, but I suspect that it isn't what most people are interested in. What if this paragraph was just replaced with something like "Contracts can be constructed using the stdlib functions `std.contract.from_predicate`, `std.contract.from_validator`, or `std.contract.custom`."

And then the "Custom contracts" section below could be replaced by

# Legacy contracts

Old versions of Nickel allowed functions of type `Label -> Dyn -> Dyn` to be treated as contracts. While this is still supported for backwards compatibility, it is discouraged and will be deprecated in the future. Use the contract constructors in `std.contract` instead.

Reply:
Right, the documentation was clumsy. In the end, the nature of contracts that can be given to `apply` are already detailed in the following section (`# Arguments`), so I just got rid of this paragraph and expanded slightly the arguments section. I trimmed the `# Custom contracts` and changed it to legacy contract as well: https://github.com/tweag/nickel/pull/1987/commits/01ca07a3e3fdc969e22c8d59166cfddd08d78522